lib/Dialect/FIRRTL/Transforms/IMDCE.cpp

```diff
@@ -0,0 +1,412 @@
+    // from the body to this instance result's SSA value, so remember it.
+    BlockArgument modulePortVal = fModule.getArgument(resultNo);
+
+    // Mark don't touch ports as alive.
+    if (hasDontTouch(modulePortVal)) {
+      markAlive(modulePortVal);
+      markAlive(instancePortVal);
+    }
+
+    resultPortToInstanceResultMapping[modulePortVal].push_back(instancePortVal);
+  }
+}
+
+void IMDCEPass::markBlockExecutable(Block *block) {
+  if (!executableBlocks.insert(block).second)
+    return; // Already executable.
+
+  for (auto &op : *block) {
+    if (isWireOrRegOrNode(&op))
+      markWireOrRegOrNode(&op);
+    else if (auto instance = dyn_cast<InstanceOp>(op))
+      markInstanceOp(instance);
+    else if (isa<FConnectLike>(op))
+      // Skip connect op.
+      continue;
+    else if (auto mem = dyn_cast<MemOp>(op))
+      markMemOp(mem);
+    else if (!mlir::MemoryEffectOpInterface::hasNoEffect(&op))
+      markUnknownSideEffectOp(&op);
+  }
+}
+
+void IMDCEPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "===----- Remove unused ports -----==="
+                          << "\n");
+  auto circuit = getOperation();
+  instanceGraph = &getAnalysis<InstanceGraph>();
+  for (auto module : circuit.getBody()->getOps<FModuleOp>()) {
+    if (module.isPublic()) {
+      markBlockExecutable(module.getBody());
+      // Mark the ports of public modules as alive.
+      for (auto port : module.getBody()->getArguments())
+        markAlive(port);
+    }
+  }
+
+  // If a value changed liveness then propagate liveness through its users and
+  // definition.
+  while (!worklist.empty()) {
+    Value changedVal = worklist.pop_back_val();
+    visitValue(changedVal);
+    for (Operation *user : changedVal.getUsers())
+      visitUser(user);
+  }
+
+  // Rewrite module signatures.
+  for (auto module : circuit.getBody()->getOps<FModuleOp>())
+    rewriteModuleSignature(module);
+
+  // Rewrite module bodies parallelly.
+  mlir::parallelForEach(circuit.getContext(),
+                        circuit.getBody()->getOps<FModuleOp>(),
+                        [&](auto op) { rewriteModuleBody(op); });
+}
+
+void IMDCEPass::visitValue(Value value) {
+  assert(isKnownAlive(value) && "only alive values reach here");
+
+  // Driving input ports propagates the liveness to each instance.
+  if (auto blockArg = value.dyn_cast<BlockArgument>()) {
+    auto module = cast<FModuleOp>(blockArg.getParentBlock()->getParentOp());
+    auto portDirection = module.getPortDirection(blockArg.getArgNumber());
+    // If the port is input, it's necessary to mark corresponding input ports of
+    // instances as alive. We don't have to propagate the liveness of output
+    // ports.
+    if (portDirection == Direction::In)
+      for (auto userOfResultPort : resultPortToInstanceResultMapping[blockArg])
+        markAlive(userOfResultPort);
+    return;
+  }
+
+  // Driving an instance argument port drives the corresponding argument of the
+  // referenced module.
```

Comment:
comment needs updating, maybe "Marking an instance port alive propagates to the corresponding port of the module"